### PR TITLE
CMCL-1144: Refactor upgrader to iterate scenes less and some bugfixes

### PR DESCRIPTION
**Pull request by gaborkb**
### Purpose of this PR

https://jira.unity3d.com/browse/CMCL-1144:
- Refactor: Upgrader opens scenes less (from 5 to 3)!
- Fixed: `DollyTracks` upgraded to `SplineDolly` could lose their `path` references.
     1. Prefab assets - fixed (by cleaning up at the end, so that references are not lost)
     2. Prefab instances - fixed (by ordering upgrades to always upgrade referencables (e.g. cmPath) first and adding this reference to the converted copy)
     3. Non-prefabs - fixed (by ordering upgrades to always upgrade referencables (e.g. cmPath))
- Fixed: `2DZoom scene` - why is the new `CmCamera` disabled. It was disabled because the gameobject has a `Zoom script` that requires `CinemachineVirtualCamera` to be present.  Vcam's `OnEnable` function detects duplicates and disabled `CmCamera`. We cannot remove `CinemachineVirtualCamera` from the gameobject because of the `RequireComponent attribute`, so it gets disabled and replaced by an enabled `CmCamera`. 
This is also the case, if the script has `[RequireComponent(typeof(CmCamera))]`.
- Error messages in console:
-> Fixed: Our animation track reference updater was too greedy - it overwrote non cm references too! 
(Warning message about setting animation track curves)

### Testing status

- [ ] Added an automated test
- [x] Passed all automated tests
- [x] Manually tested above mentioned fixes

### Documentation status

- [ ] Updated [CHANGELOG](https://keepachangelog.com/en/1.0.0/)
- [ ] Updated README (if applicable)
- [ ] Commented all public classes, properties, and methods
- [ ] Updated user documentation

### CODE REVIEW

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs [hunk 1]
```diff
@@ -86,8 +86,18 @@ public GameObject UpgradeComponents(GameObject go)
                      orbital.UpgradeToCm3(go.GetComponent<CinemachineOrbitalFollow>());
                      ConvertInputAxis(go, "Look Orbit X", ref orbital.m_XAxis, ref orbital.m_RecenterToTargetHeading);
                 }
+
                 if (ReplaceComponent<CinemachineTrackedDolly, CinemachineSplineDolly>(go))
-                    go.GetComponent<CinemachineTrackedDolly>().UpgradeToCm3(go.GetComponent<CinemachineSplineDolly>());
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:90]: Why was this not covered by the ordering of RootUpgradeComponentTypes?  Paths are supposed to get converted first.  What is the usecase that shows this problem?

I don't like this fix because it secretly upgrades 2 objects when this class is supposed to upgrade one object at a time.

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineClearShot.cs [hunk 2]
```diff
@@ -85,7 +85,7 @@ protected internal override void LegacyUpgrade(int streamedVersion)
             {
                 DefaultTarget = new DefaultTargetSettings 
                 { 
-                    Enabled = m_LegacyLookAt != null || m_LegacyFollow != null,
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineClearShot.cs:88]: Why this change?

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs [hunk 3]
```diff
@@ -131,7 +131,7 @@ protected internal override void LegacyUpgrade(int streamedVersion)
             {
                 DefaultTarget = new DefaultTargetSettings 
                 { 
-                    Enabled = m_LegacyLookAt != null || m_LegacyFollow != null,
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: Why this change?

File: com.unity.cinemachine/Runtime/Deprecated/CinemachineTrackedDolly.cs [hunk 4]
```diff
@@ -349,8 +349,6 @@ internal void UpgradeToCm3(CinemachineSplineDolly c)
                 case CinemachinePathBase.PositionUnits.Normalized: c.PositionUnits = UnityEngine.Splines.PathIndexUnit.Normalized; break;
             }
             c.SplineOffset = m_PathOffset;
-            if (m_Path != null)
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Deprecated/CinemachineTrackedDolly.cs:352]: Why this change?  Paths should have been upgraded first

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs [hunk 5]
```diff
@@ -345,34 +179,72 @@ void UpgradePrefabsAndPrefabInstances()
...
+        void UpgradeAllScenes(Dictionary<int, List<ConversionLink>> conversionLinksPerScene)
+        {
             for (var s = 0; s < m_SceneManager.SceneCount; ++s)
             {
+                // 0. Open scene
                 var scene = OpenScene(s);
                 var timelineManager = new TimelineManager(scene);
-                var conversionLinks = conversionLinksPerScene[m_SceneManager.GetScenePath(s)];
-                var allGameObjectsInScene = GetAllGameObjects();
+                var upgradables = GetUpgradeCandidates(scene.GetRootGameObjects());
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:202]: I think here is the answer to my other question.  The correct algorithm might have to be to do this whols thing in 2 passes: 
1. All the leaf objects that may be referenced (e.g. paths)
2. All the things that might refer to other upgraded things (e.g. vcams, freeLooks)

If in fact that is the solution, then we will lose the optimization because we'll have to double the number of scene opens

**Review by glabute — CHANGES_REQUESTED** [review]

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineClearShot.cs (hunk 2 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineClearShot.cs:88]: Fixes: [CMCL-1157](https://jira.unity3d.com/browse/CMCL-1157) - we may need to come up with something else. LegacyUpgrade cannot check for null, because that is only allowed on the main thread. However, OnAfterDeserialize is not guaranteed to run on the main thread. When it does not run on main thread we get an error that stops LegacyUpgrade.
(Error message about EnsureRunningOnMainThread)

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs (hunk 3 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: Fixes: [CMCL-1157](https://jira.unity3d.com/browse/CMCL-1157) - we may need to come up with something else. LegacyUpgrade cannot check for null, because that is only allowed on the main thread. However, OnAfterDeserialize is not guaranteed to run on the main thread. When it does not run on main thread we get an error that stops LegacyUpgrade.
(Error message about EnsureRunningOnMainThread)

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs (hunk 5 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:202]: I'll take a look

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs (hunk 3 above)
Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: I am confused.  What error was happening, and when was it happening?  The Jira ticket doesn't say

Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: The error happens when LegacyUpgrade is **not** called from the main thread. 

OnAfterDeserialize calls LegacyUpgrade. Unfortunately, OnAfterDeserialize is not guaranteed to be called from the main thread.

Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: Can you be more specific?  Where is the offending null check?

Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: nvm I understand now :)

Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs:134]: Let's pull this out into a separate PR.  I think we need a different fix.

Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineStateDrivenCamera.cs]: https://github.com/Unity-Technologies/com.unity.cinemachine/pull/534

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.Data.cs [hunk 6]
```diff
@@ -21,7 +21,15 @@ partial class UpgradeObjectToCm3
...
             typeof(CinemachineVirtualCamera),
         };
-        
+
+        /// <summary>
+        /// Any component that may be referenced by vcams or freelooks 
+        /// </summary>
+        public static readonly List<Type> Referencables = new()
+        {
+            typeof(CinemachinePath)
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.Data.cs:30]: What about inemachineSmoothPath?

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs [hunk 7]
```diff
@@ -113,204 +113,47 @@ public static void UpgradeProject()
...
-                foreach (var director in directors)
+                var upgradables = 
+                    GetUpgradables(scene.GetRootGameObjects(), UpgradeObjectToCm3.Referencables, false);
+                foreach (var upgradable in upgradables)
                 {
-                    if (renames.ContainsKey(director.name)) // search based on guid name
+                    if (PrefabUtility.IsPartOfAnyPrefab(upgradable)) 
+                        continue;
+                    
+                    foreach (var referencable in UpgradeObjectToCm3.Referencables)
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:137]: This is weird.  Why are you looping again?  If upgradable contains multiple upgradable components, UpgradeObjectComponents() will handle it.  No need to upgrade multiple times.

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs [hunk 8]
```diff
@@ -370,13 +220,51 @@ void UpgradePrefabsAndPrefabInstances()
...
+        static List<GameObject> GetAllGameObjects()
+        {
+            var all = (GameObject[])Resources.FindObjectsOfTypeAll(typeof(GameObject));
+            return all.Where(go => !EditorUtility.IsPersistent(go.transform.root.gameObject)
+                && (go.hideFlags & (HideFlags.NotEditable | HideFlags.HideAndDontSave)) == 0).ToList();
+        }
+
+        static void RestoreConvertedCopyReferences(GameObject original, GameObject converted, List<ConversionLink> conversionLinks)
+        {
+            // TODO: if we'll have more components that can reference upgradables, then we'll need to
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:240]: Big red hack-flags here.  And what about dolly cart?

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs [hunk 9]
```diff
@@ -370,13 +220,51 @@ void UpgradePrefabsAndPrefabInstances()
...
+            // TODO: create some Data in UpgradeObjectToCM3.Data to make this generic and loopable
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:241]: I think if we have some general reflection-based code to handle references in any script, this problem will solve itself

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs [hunk 10]
```diff
@@ -400,59 +290,98 @@ void UpgradePrefabsAndPrefabInstances()
                             continue; // is a backup copy
 
                         UpgradeObjectComponents(c.gameObject, timelineManager);
+                    }
+
+                    // clean-up now, because we needed obsolete components to restore references
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:295]: I'm confused by this.  When did we restore references?

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs [hunk 11]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Linq;
 using Cinemachine.Utility;
+using NUnit.Framework;
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:7]: NUnit?  no unit tests here.  We should not have this dependancy

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs [hunk 12]
```diff
@@ -86,8 +87,17 @@ public GameObject UpgradeComponents(GameObject go)
...
                 if (ReplaceComponent<CinemachineTrackedDolly, CinemachineSplineDolly>(go))
-                    go.GetComponent<CinemachineTrackedDolly>().UpgradeToCm3(go.GetComponent<CinemachineSplineDolly>());
+                {
+                    var obsoleteDolly = go.GetComponent<CinemachineTrackedDolly>();
+                    var splineDolly = go.GetComponent<CinemachineSplineDolly>();
+                    obsoleteDolly.UpgradeToCm3(splineDolly);
+                    
+                    // set splineDolly spline reference
+                    if (obsoleteDolly.m_Path != null) 
+                        obsoleteDolly.m_Path.TryGetComponent(out splineDolly.Spline);
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:99]: Should do it in obsoleteDolly.UpgradeToCm3(), same way that dollycart does it, above

**Review by glabute — CHANGES_REQUESTED** [review]

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.Data.cs (hunk 6 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.Data.cs:30]: Yes, it was meant to be the base class -> CinemachinePathBase.

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs (hunk 7 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:137]: Yes. Sorry! Good catch! 👍  
The reason why I refactored GetUpgradables was to remove this part. Originally, I had a quick hack here just to filter out anything that's not referable.

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs (hunk 8 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:240]: Definitely. I won't remove this for now, because the reflection based referencing will obsolete this anyway. Until that I leave the hack here.

File: com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs (hunk 10 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.cs:295]: In UpgradeObjectComponents, in the foreach above this foreach

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs (hunk 11 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:7]: I am using assert in the switch statement to throw an error if we ever have a new path type -> unlikely. I could remove the assert or add a throw new NotImplementedException.

Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:7]: The user may have a special path. So maybe throwing the not implemented exception is good. Or we could just ignore those paths.

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs (hunk 12 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:99]: Yes

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs (hunk 11 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs]: Tested. If user has a custom CinemachinePathBase, then we add a Spline (so that the upgraded trackedDolly can reference it) on it and display error in console saying that their custom thing was not updated. The gameobject also gets the CinemachineDoNotUpgrade tag.

Is this the desired behaviour? I am not sure if the `CinemachineDoNotUpgrade` tag is necessary.

**Review by glabute — APPROVED** [review]

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs (hunk 11 above)
Comment (glabute) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs]: I guess it's good enough for now

File: com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs [hunk 13]
```diff
@@ -182,6 +189,8 @@ static int FindRig(string path)
         /// <param name="go">The GameObject being upgraded</param>
         public void DeleteObsoleteComponents(GameObject go)
         {
+            if (go == null) return;
```
[the remark this answers is not in the document]
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:192]: I think I was testing without deleting any components and I got the error here because go was null. Not something that should happen in a real situtation. It does not hurt having it here, but I can remove it.

[the remark this answers is not in the document]
Comment (glabute) [reply com.unity.cinemachine/Editor/Upgrader/UpgradeObjectToCm3.cs:192]: It doesn't hurt, but it can hide potential bugs.  It's part of general code cleanliness: better to crash early on an unexpected null than to try to find the source of a covered-up problem
